[CI] CrossClusterAsyncEnrichStopIT testEnrichAfterStop failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #5990 / oraclelinux-7_platform-support-unix](https://gradle-enterprise.elastic.co/s/ajegktoml5opi)
- [elasticsearch-periodic-platform-support #5990 / oraclelinux-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/a2ee34bfab3xe)
- [elasticsearch-periodic-platform-support #5990 / sles-12_platform-support-unix](https://gradle-enterprise.elastic.co/s/fqjvieix2fbl6)
- [elasticsearch-periodic-platform-support #5990 / almalinux-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/jqaom3bhspz2i)
- [elasticsearch-periodic-platform-support #5990 / ubuntu-2204_platform-support-unix](https://gradle-enterprise.elastic.co/s/ggq2holdl457q)
- [elasticsearch-periodic-platform-support #5990 / debian-12_platform-support-unix](https://gradle-enterprise.elastic.co/s/2stpd26d2pejk)
- [elasticsearch-periodic-platform-support #5990 / amazonlinux-2023_platform-support-aws](https://gradle-enterprise.elastic.co/s/yduq552r3d64m)
- [elasticsearch-periodic-platform-support #5990 / ubuntu-2404-aarch64_checkpart3_platform-support-arm](https://gradle-enterprise.elastic.co/s/czocuydauhoda)
- [elasticsearch-periodic-platform-support #5990 / ubuntu-2004-aarch64_checkpart3_platform-support-arm](https://gradle-enterprise.elastic.co/s/c2nayp3yzne3k)
- [elasticsearch-periodic-platform-support #5985 / almalinux-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/k3hmxyonemx7k)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:internalClusterTest" --tests "org.elasticsearch.xpack.esql.action.CrossClusterAsyncEnrichStopIT.testEnrichAfterStop" -Dtests.seed=F0133A9CF18832C7 -Dtests.locale=luy-Latn-KE -Dtests.timezone=Africa/Gaborone -Druntime.java=23
```

**Applicable branches:**
8.x

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.action.CrossClusterAsyncEnrichStopIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testEnrichAfterStop),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: <[[5, Apple], [6, Microsoft], [3, Redhat], [2, Samsung], [2, null]]>
     but: was <[[3, Microsoft], [1, Samsung], [1, null]]>
```

**Issue Reasons:**
- [8.x] 7 consecutive failures in test testEnrichAfterStop
- [8.x] 2 consecutive failures in step sles-12_platform-support-unix
- [8.x] 2 consecutive failures in step almalinux-9_platform-support-unix
- [8.x] 3 consecutive failures in step ubuntu-2204_platform-support-unix
- [8.x] 2 consecutive failures in step amazonlinux-2023_platform-support-aws
- [8.x] 25 failures in test testEnrichAfterStop (18.9% fail rate in 132 executions)
- [8.x] 2 failures in step oraclelinux-9_platform-support-unix (66.7% fail rate in 3 executions)
- [8.x] 2 failures in step sles-12_platform-support-unix (66.7% fail rate in 3 executions)
- [8.x] 2 failures in step almalinux-9_platform-support-unix (66.7% fail rate in 3 executions)
- [8.x] 3 failures in step ubuntu-2204_platform-support-unix (100.0% fail rate in 3 executions)
- [8.x] 2 failures in step amazonlinux-2023_platform-support-aws (66.7% fail rate in 3 executions)
- [8.x] 3 failures in pipeline elasticsearch-periodic-platform-support (100.0% fail rate in 3 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -209,9 +209,6 @@ tests:
   issue: https://github.com/elastic/elasticsearch/issues/120668
 - class: org.elasticsearch.xpack.security.authc.ldap.ADLdapUserSearchSessionFactoryTests
   issue: https://github.com/elastic/elasticsearch/issues/119882
-- class: org.elasticsearch.xpack.esql.action.CrossClusterAsyncEnrichStopIT
-  method: testEnrichAfterStop
-  issue: https://github.com/elastic/elasticsearch/issues/120757
 - class: org.elasticsearch.xpack.test.rest.XPackRestIT
   method: test {p0=ml/3rd_party_deployment/Test start deployment fails while model download in progress}
   issue: https://github.com/elastic/elasticsearch/issues/120810

--- a/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeBuffer.java
+++ b/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeBuffer.java
@@ -47,7 +47,17 @@ final class ExchangeBuffer {
             notifyNotEmpty();
         }
         if (noMoreInputs) {
-            discardPages();
+            // O(N) but acceptable because it only occurs with the stop API, and the queue size should be very small.
+            if (queue.removeIf(p -> p == page)) {
+                page.releaseBlocks();
+                final int size = queueSize.decrementAndGet();
+                if (size == maxSize - 1) {
+                    notifyNotFull();
+                }
+                if (size == 0) {
+                    completionFuture.onResponse(null);
+                }
+            }
         }
     }
 

--- a/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/exchange/ExchangeBufferTests.java
+++ b/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/exchange/ExchangeBufferTests.java
@@ -66,6 +66,25 @@ public class ExchangeBufferTests extends ESTestCase {
         blockFactory.ensureAllBlocksAreReleased();
     }
 
+    public void testOutstandingPages() throws Exception {
+        ExchangeBuffer buffer = new ExchangeBuffer(randomIntBetween(1000, 10000));
+        var blockFactory = blockFactory();
+        Page p1 = randomPage(blockFactory);
+        Page p2 = randomPage(blockFactory);
+        buffer.addPage(p1);
+        buffer.addPage(p2);
+        buffer.finish(false);
+        buffer.addPage(randomPage(blockFactory));
+        assertThat(buffer.size(), equalTo(2));
+        assertSame(buffer.pollPage(), p1);
+        p1.releaseBlocks();
+        assertSame(buffer.pollPage(), p2);
+        p2.releaseBlocks();
+        assertNull(buffer.pollPage());
+        assertTrue(buffer.isFinished());
+        blockFactory.ensureAllBlocksAreReleased();
+    }
+
     private static MockBlockFactory blockFactory() {
         BigArrays bigArrays = new MockBigArrays(PageCacheRecycler.NON_RECYCLING_INSTANCE, ByteSizeValue.ofGb(1)).withCircuitBreaking();
         CircuitBreaker breaker = bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST);

